# error

Una herramienta de [Sumor Cloud](https://sumor.cloud).  
[Más Documentación](https://sumor.cloud)
Una biblioteca de manejo de errores con soporte multi-idioma, código de error predefinido y paso de datos.

[![CI](https://github.com/sumor-cloud/error/actions/workflows/ci.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ci.yml)
[![Test](https://github.com/sumor-cloud/error/actions/workflows/ut.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ut.yml)
[![Coverage](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml)
[![Audit](https://github.com/sumor-cloud/error/actions/workflows/audit.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/audit.yml)

## Instalación

```bash
npm i @sumor/error --save
```

## Prerrequisitos

### Versión de Node.JS

Requiere Node.JS versión 16.x o superior

### Requerir módulo ES de Node.JS

Como este paquete está escrito en módulo ES,
por favor cambie el siguiente código en su archivo `package.json`:

```json
{
  "type": "module"
}
```

## Uso

### Definir Error

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'Usuario no encontrado',
    USER_EXISTED: 'Usuario {name} existente'
  }
})

throw new MyError('USER_NOT_FOUND')
// output: Error: Usuario no encontrado

throw new MyError('USER_EXISTED', { name: 'Alice' })
// output: Error: Usuario Alice existente
```

### Multi-idioma

```js
import defineError from '@sumor/error'

const MyError = defineError({
  language: 'es', // idioma por defecto
  code: {
    USER_NOT_FOUND: 'Usuario no encontrado',
    USER_EXISTED: 'Usuario {name} existente'
  },
  i18n: {
    en: {
      USER_NOT_FOUND: 'User not found',
      USER_EXISTED: 'User {name} existed'
    }
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
error.language = 'es' // cambiar idioma del error
console.log(error)
// output: Error: Usuario Alice existente

error.language = 'en' // cambiar idioma del error
console.log(error)
// output: Error: User Alice existed
```

### Convertir Error a JSON

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'Usuario no encontrado',
    USER_EXISTED: 'Usuario {name} existente'
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
console.log(error.json())
// output: {"code":"USER_EXISTED","message":"Usuario Alice existente"}
```

### Error Subyacente

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'Fallo en la verificación del campo',
    FIELD_CANNOT_EMPTY: 'El campo {name} no puede estar vacío',
    FIELD_TOO_LONG: 'El campo {name} es demasiado largo'
  },
  i18n: {
    en: {
      FIELD_VERIFY_FAILED: 'Field verify failed',
      FIELD_CANNOT_EMPTY: 'Field {name} cannot be empty',
      FIELD_TOO_LONG: 'Field {name} is too long'
    }
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'usuario' }),
  new MyError('FIELD_TOO_LONG', { name: 'contraseña' })
])

console.log(error.json())
/* 
output: 
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Fallo en la verificación del campo",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"El campo usuario no puede estar vacío"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"El campo contraseña es demasiado largo"
    }
  ]
}
*/

error.language = 'en'
console.log(error.json())
/*
output:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Field verify failed",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"Field username cannot be empty"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"Field password is too long"
    }
  ]
}
*/
```

### Combinar Error Estándar

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'Fallo en la verificación del campo',
    FIELD_CANNOT_EMPTY: 'El campo {name} no puede estar vacío',
    FIELD_TOO_LONG: 'El campo {name} es demasiado largo'
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'usuario' }),
  new MyError('FIELD_TOO_LONG', { name: 'contraseña' }),
  new Error('Error Desconocido')
])

console.log(error.json())
/*
output:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Fallo en la verificación del campo",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"El campo usuario no puede estar vacío"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"El campo contraseña es demasiado largo"
    },{
      "code":"UNKNOWN_ERROR",
      "message":"Error Desconocido"
    }
  ]
}
 */
```
